CREATE INDEX now saves a new index and its properties in the catalog

saketj opened the pull request:
This is the final PR towards CREATE INDEX implementation in Quickstep that saves newly created indices and their properties in the catalog. 
The CatalogProto has been modified to define a new message type- 'IndexScheme', which contains the set of indices defined on the relation. Within the index scheme, each index is represented as an IndexEntry that contains a string index_name, and an associated protobuf IndexSubBlockDescription. 

Addition of an index creates an IndexEntry in the catalog and also adds a new index sub block to the StorageBlockLayout. Duplicate indices can be checked by going through a map of indices maintained by the catalog. The PR also adds support for flexible parsing and definition of any number of properties on a given index.

zuyu commented on catalog/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ add_library(quickstep_catalog_CatalogRelationSchema
             CatalogRelationSchema.hpp)
 add_library(quickstep_catalog_CatalogTypedefs ../empty_src.cpp CatalogTypedefs.hpp)
 add_library(quickstep_catalog_PartitionScheme PartitionScheme.cpp PartitionScheme.hpp)
+add_library(quickstep_catalog_IndexScheme IndexScheme.cpp IndexScheme.hpp)
```
Please move this line between `quickstep_catalog_CatalogTypedefs` and `quickstep_catalog_PartitionScheme`, as we sort them in alphabet order.

Note that we should also reorder `quickstep_catalog_NUMAPlacementScheme` to the right place in another PR.

saketj replied:
Fixed.

zuyu commented on catalog/Catalog.proto:
```diff
@@ -57,6 +57,16 @@ message NUMAPlacementScheme {
+    required string index_name = 1;
+    repeated IndexSubBlockDescription index_description = 2;
+  }
+  repeated IndexEntry index_entries = 2;
```
Does `num_indices()` equal to `index_entries_size()`?

If so, we don't need the field for `num_indices`.

saketj replied:
Removed num_indices from the proto definition.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -93,9 +93,9 @@ CatalogRelation::CatalogRelation(const serialization::CatalogRelation &proto)
-    indices_.push_back(proto.index(i));
+  // Deserializing the index scheme defined for the relation, if any
+  if (proto.has_index_scheme()) {
+    setIndexScheme(IndexScheme::DeserializeIndexScheme(proto.index_scheme()));
```
Rename `DeserializeIndexScheme` to `ReconstructFromProto` for function naming consistency.

saketj replied:
Done.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -170,6 +171,44 @@ class CatalogRelation : public CatalogRelationSchema {
+   *          is first created.
+   *
+   * @param index_scheme The index scheme object for a relation, which
+   *        becomes owned by this relation.
```
Add the following explicit comment for the ownership:

```
@note setIndexScheme takes ownership of \c index_scheme.
```

saketj replied:
Added.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -170,6 +171,44 @@ class CatalogRelation : public CatalogRelationSchema {
+   * @return True if the relation has a index scheme, false otherwise.
+   **/
+  bool hasIndexScheme() const {
+    return index_scheme_.get() != nullptr;
```
`index_scheme_ != nullptr` is fine.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
-      return true;
-    } else {
+  bool hasIndexWithName(const std::string &index_name) const {
+    if (index_scheme_.get() == nullptr) {
```
`index_scheme_ == nullptr` is fine.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
+    if (index_scheme_.get() == nullptr) {
       return false;
+    } else {
+      return index_scheme_->hasIndexWithName(index_name);
```
Refactor this function to

```
return index_scheme_ && index_scheme_->hasIndexWithName(index_name);
```

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
+      return false;
+    } else {
+      return index_scheme_->hasIndexWithDescription(index_description);
+    }
```
Refactor this function to

```
return index_scheme_
    && index_scheme_->hasIndexWithDescription(index_description);
```

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
+    }
+
+    // verify that index with similar description does not exist
+    std::vector<IndexSubBlockDescription>::const_iterator itr;
```
Remove this line, and use `auto` in the for-loop.

Btw, we typically use `it` when shortening `iterator`, instead of `itr`, while `cit` for `const_iterator`.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
-    indices_.emplace_back(name);
+  bool addIndex(const std::string &index_name, const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    // create an index_scheme, if it does not exist
+    if (index_scheme_.get() == nullptr) {
```
Remove `get()`.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
+          std::size_t num_indicies = layout->index_description_size();
+          // update the description of newly added index
+          // whose (position = num_indices - 1)
+          layout->mutable_index_description(num_indicies - 1)->CopyFrom(*itr);
```
I think the following refactor should work:

```
layout->add_index_description()->MergeFrom(*itr);
```

The reason why we use `MergeFrom`, instead of `CopyFrom`, is for a better performance. We don't need to free up the space for the newly-added index description.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,204 @@
+   * @param attr_type The attribute type of the partitioning attribute.
+   * @return The deserialied partition scheme object.
+   **/
+  static IndexScheme* DeserializeIndexScheme(const serialization::IndexScheme &proto);
```
As commented above, use `ReconstructFromProto` instead.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+      return false;
+    }
+    // Check if the number of indices is non-zero.
+    if (proto.num_indices() <= 0) {
```
As commented above, no needs for `num_indices`.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+    if (proto.num_indices() <= 0) {
+      return false;
+    }
+    return true;
```
Also check whether all `IndexSubBlockDescription`s are valid, if any.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+    // set the entries of the index scheme
+    std::size_t num_indices = proto.num_indices();
+    std::size_t index_number = 0;
+    std::unordered_map<std::string, std::vector<IndexSubBlockDescription>>::const_iterator itr;
```
Remove this line, and use `auto` in the for-loop.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+
+      // populate the details of the index entry
+      index_entry->set_index_name(itr->first);
+      std::vector<IndexSubBlockDescription> index_descriptions = itr->second;
```
Include `<utility>` as we use `first` and `second`.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+      DEBUG_ASSERT(index_number < num_indices);
+      // create an index entry
+      proto.add_index_entries();
+      serialization::IndexScheme_IndexEntry *index_entry = proto.mutable_index_entries(index_number);
```
Remove `index_number`, and refactor the above code to

```
serialization::IndexScheme_IndexEntry *index_entry = proto.add_index_entries();
```

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+      index_entry->set_index_name(itr->first);
+      std::vector<IndexSubBlockDescription> index_descriptions = itr->second;
+      std::size_t index_description_num = 0;
+      std::vector<IndexSubBlockDescription>::iterator index_description_itr;
```
Remove the above two lines, and use `auto` for the iterator.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,104 @@
+        // add each index description to the index entry
+        index_entry->add_index_description();
+        index_entry->mutable_index_description(index_description_num)->CopyFrom(*index_description_itr);
+        ++index_description_num;
```
Refactor the above codes to

```
index_entry->add_index_description()->MergeFrom(*index_description_itr);
```

saketj replied:
Refactored to use the mutable description returned by add_index_description() in all instances.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,204 @@
+      return true;
+    } else {
+      return false;
+    }
```
Just refactor to:

```
return index_map_.find(index_name) != index_map_.end();
```

saketj replied:
Refactored.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,204 @@
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map
+    std::unordered_map<std::string, std::vector<IndexSubBlockDescription>>::const_iterator itr;
```
Use `auto`.

saketj replied:
Replaced iterator types with auto references.

saketj commented:
@zuyu Thanks Zuyu, for the detailed review and the comments on the PR. These were really helpful. I have now fixed these issues in the latest commit. 
I had one quick question on the IWYU script that is there in the third party folder. It seems that when I ran it on `catalog/IndexScheme.hpp`, it asked me to remove unordered_map, which I am using for sure.
Following was the command that I ran
`./third_party/iwyu/iwyu_helper.py catalog/IndexScheme.hpp`
Should I be running some other IWYU script? or, should I interpret the output with caution?

cramja commented:
@saketj Regarding IWYU, it's not perfect and will suggest incorrect things.

Here's what @zuyu had to say about it when I asked that question:

> @cramja No, IWYU just provides some suggestions, but also introduces false positive. Use your own judgements to add the headers for new changes. There are many legacy IWYU issues to fix in a separate PR.

saketj commented:
@cramja Thanks Marc, that explains it. Sure, I will interpret the IWYU results with caution then.

zuyu commented:
@saketj There are conflicts to fix, as the master branch moves forward.

cramja commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,59 @@ void ExecutionGenerator::convertCreateIndex(
       optimizer_context_->catalog_database()->getRelationByIdMutable(
             input_relation_info->relation->getID());
+
+  // check if any index with the specified name already exists
```
Generally comments should be sentences (Sentences have a capitalized first word and a period following.)

cramja replied:
By the way, hello! I am using this PR in a benchmark with an index so while it is compiling, I'll parse your code.

saketj replied:
Addressed.

cramja commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,59 @@ void ExecutionGenerator::convertCreateIndex(
+  }
+
+  // assertion: must have at least one attribute specified to build index upon
+  DEBUG_ASSERT(physical_plan->index_attributes().size() > 0);
```
`DEBUG_ASSERT` is deprecated. [Google logging](https://google-glog.googlecode.com/svn/trunk/doc/glog.html) is preferred.

saketj replied:
Fixed.

cramja commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,59 @@ void ExecutionGenerator::convertCreateIndex(
+  for (const E::AttributeReferencePtr &attribute : physical_plan->index_attributes()) {
+    const CatalogAttribute *catalog_attribute
+      = input_relation->getAttributeByName(attribute->attribute_name());
+    DEBUG_ASSERT(catalog_attribute != nullptr);
```
`DEBUG_ASSERT`-> glog

cramja replied:
Would it make sense to throw an SQLError here with an error message, "The specified attribute does not exist in relation *foo*"?

saketj replied:
@cramja Marc, the Resolver::resolveCreateIndex() does this very check and prints the attribute undefined error message. So, I thought I could skip it here.

saketj replied:
Replaced DEBUG_ASSERT with DCHECK.

cramja commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,59 @@ void ExecutionGenerator::convertCreateIndex(
+      index_descriptions.emplace_back(index_description);
+    }
+  }
+
```
It would be good to check that the `DescriptionIsValid` at this point, similar to how [BlockProperties](https://github.com/pivotalsoftware/quickstep/blob/master/query_optimizer/ExecutionGenerator.cpp#L800) does it.

cramja replied:
It sort of acts as a catch-all for any errors which the Resolver could not capture.

cramja replied:
Though I see you have a variable `is_description_valid_` in the Parser classes, maybe this will catch problems much earlier in the translation process.

cramja replied:
Though actually, it would still be a good safe-guard because making an invalid description will cause QS to crash.

saketj replied:
Thanks. Have added this check now.

cramja commented on parser/ParseStatement.hpp:
```diff
@@ -195,22 +197,55 @@ class ParseStatementCreateIndex : public ParseStatement {
+                              ParseString *relation_name,
+                              PtrList<ParseAttribute> *attribute_list,
+                              ParseString *index_type,
+                              const int index_properties_line_number,
```
As a suggestion, it might make it easier to resolve index properties if they were abstracted into a small helper class, `IndexProperties` which would essentially be a wrapper around the properties. It could do things like check for duplicate properties, or lookup a property with a given key. 

I found it to be helpful to do it this way in the `BlockProperties` feature

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+ */
+class IndexProperties {
+ public:
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_t;
```
I would guess that the `_t` postfix is not necessary or should be changed to a full spelling, `valid_property_map_type` Grepping through the codebase finds that most of the typedefs that are not auto-generated are of the form `_type` or have no postfix

saketj replied:
Replaced

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+/**
+ * @brief Base class for different types of indices and their associated properties
+ */
+class IndexProperties {
```
Good idea to use this abstraction to wrap the list of properties!

cramja replied:
Oh, I see, these aren't necessarily Parser classes, they're more like resolver helpers. Maybe they should go in `query_optimizer/resolver/`

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+ */
+class CSBTreeIndexProperties : public IndexProperties {
+ public:
+  CSBTreeIndexProperties() {
```
Most of these should be set in the initializer list similar to:
```c++
class Foo {
  Foo(int a) : a_(a) { }
 private:
  int a_;
};
```
This is because c++ will have already called default initializers by the time you enter the code block which is extra work and might cause unexpected behavior.

zuyu replied:
You may want to add `public:` before the constructor.

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+class CSBTreeIndexProperties : public IndexProperties {
+ public:
+  CSBTreeIndexProperties() {
+    is_index_description_valid_ = true;
```
How is `is_index_description_valid_` this going to be determined?

zuyu replied:
Use the constructor init list instead.

saketj replied:
@cramja Every supported index starts with an initialized true value for the is_index_description_valid_ (Bloom filter is not yet supported, hence has a default false) value. Later, if custom properties are specified then this flag is modified depending on whether the parse succeeded or not.

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+    invalid_property_node_ = nullptr;
+  }
+
+  void setIndexDescriptionAsInvalid(const ParseKeyValue *invalid_node, std::string reason) {
```
You could use a default argument to avoid repeated boilerplate:
```c++
void setIndexDescriptionAsInvalid(
  std::string reason, 
  const ParseKeyValue *invalid_node = nullptr) { ... }
```

zuyu replied:
Why make a deep copy of `reason`, instead of using a reference?

```c++
void setIndexDescriptionAsInvalid(
  const std::string &reason,
  const ParseKeyValue *invalid_node);
```

As for the code style, we typically put the pointer argument last.

saketj replied:
@zuyu updated this now. Thanks.

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
+
+  bool hasDuplicateCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
```
Is the word 'custom' necessary in these method names? I'd suggest trimming it down to `hasDuplicateProperty` though you do not need to change it.

saketj replied:
@cramja Marc, here the 'custom' word is to being used to distinguish properties, which are specified through the CREATE INDEX statement, from the default properties of an index.

cramja commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,271 @@
+ protected:
+  valid_property_map_t valid_property_map_;  // a map of index property constants and their associated type
+  bool is_index_description_valid_;
+  const ParseKeyValue* invalid_property_node_;  // referred object owned by ParseIndexProperties class
```
Pointer should move over to the var name: `const ParseKeyValue *invalid_property_node_;`

cramja commented:
@saketj I used your code to make a CSBTree on a benchmarking study today, good work!

While I was waiting for it to compile and run, I looked over `parser/ParseIndexProperties.hpp` as well as the parts of the execution generator related to adding the indexed_attribute extension.

I like your creation of helper classes to do resolving! I'll admit that I thought the code for BlockProperties in resolver was cluttered. Even after several refactors.

Maybe this shared code for
```
+/**
+ * @brief Base class for different types of indices and their associated properties
+ */
+class IndexProperties {
```
could be abstracted into a class in the resolver folder. Doesn't have to be this PR, but as an enhancement for the next time we hire an intern (kidding, but if you know any aspiring undergrads...).

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,59 @@ void ExecutionGenerator::convertCreateIndex(
   // CreateIndex is converted to a CreateIndex operator.
   const CatalogRelationInfo *input_relation_info =
       findRelationInfoOutputByPhysical(physical_plan->input());
+  // NOTE: the object pointed to by input_relation is owned by CatalogRelation
```
No need for this comment.

saketj commented:
@zuyu  Thanks Zuyu for the heads up on the merge. I have now merged recent changes from the master.

saketj commented:
@cramja  Thanks Marc, for the detailed review. These are some great comments and I will address them all soon by tomorrow. I am glad that CREATE INDEX was useful.

cramja commented:
@saketj If moving the `IndexProperties` classes to the resolver folder is egregious, then it doesn't have to be part of this PR IMO.

zuyu commented on relational_operators/CreateIndexOperator.hpp:
```diff
@@ -54,11 +59,14 @@ class CreateIndexOperator : public RelationalOperator {
    **/
   CreateIndexOperator(CatalogRelation *relation,
-                      const std::string &index_name)
+                      const std::string index_name,
```
No need to remove the reference symbol.

saketj replied:
Fixed.

zuyu replied:
No, `index_name` still misses the reference symbol.

saketj replied:
Yeah sorry, I meant I fixed it my local branch, will push all the changes together once I am done. I usually go on marking all the suggested edits before I push them, so that I track all the ones that I have fixed.

zuyu replied:
So, you've committed a write-ahead log.

saketj replied:
Its committed now. :)

zuyu commented on relational_operators/CreateIndexOperator.hpp:
```diff
@@ -75,7 +83,8 @@ class CreateIndexOperator : public RelationalOperator {
  private:
   CatalogRelation *relation_;
-  const std::string &index_name_;
+  std::string index_name_;
```
Why made this change?

saketj replied:
@zuyu I was wondering that whether initializing a private member string with a reference should be recommended or not. Because lets say in the distributed version, what if the referenced string changes between the time the CreateIndexOperator is marked for execution by Foreman and it is actually executed? Is it okay to use a reference in such cases?

zuyu replied:
We avoid deep copies as possible for the sake of high performance.

The original code should be ok in the distributed case, and I think it is still good for concurrent queries.

Under which scenario do you think the referenced string will change? It should be fixed during the query optimization, right?

saketj replied:
@zuyu Yes, it shouldn't change during query optimization as the index_name string is part of the phyiscal plan that is generated during the query optimization. But the operator is part of the DAG that gets passed to Foreman. I am may wrong here, but is the plan kept around till the DAG is actually executed? If it is, then the references will never changes.

saketj replied:
@zuyu Looks like there is a single query plan pointer that is maintained and is passed to the Foreman, so the references will never change. Thanks for pointing this out. I will update this.

zuyu commented on storage/StorageBlockLayout.proto:
```diff
@@ -77,6 +78,15 @@ message CSBTreeIndexSubBlockDescription {
+    repeated int32 indexed_attribute_id = 33;
+    optional fixed64 bloom_filter_size = 34 [default = 10000];
+    optional int32 number_of_hashes = 35 [default = 20];
+    optional fixed64 projected_element_count = 36 [default = 1000000];
```
Make comments for above three default values. Why choose them?

saketj replied:
Done.

saketj commented:
@zuyu  is it advisable to squash git commits in an on-going pull request? I was wondering if I should rewrite the many fix commits, I have already pushed in this PR to one single commit.

zuyu commented:
@saketj Yes, please do so!

saketj commented:
Multiple small git commits to this PR have been squashed now into a single commit that fixes all the issues found so far.

zuyu commented on catalog/CMakeLists.txt:
```diff
@@ -81,6 +82,7 @@ target_link_libraries(quickstep_catalog_CatalogRelation
                       quickstep_catalog_CatalogTypedefs
                       quickstep_catalog_Catalog_proto
                       quickstep_catalog_PartitionScheme
+                      quickstep_catalog_IndexScheme
```
Move `quickstep_catalog_IndexScheme` before `quickstep_catalog_PartitionScheme`.

saketj replied:
Fixed.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -93,9 +93,9 @@ CatalogRelation::CatalogRelation(const serialization::CatalogRelation &proto)
-  // Deserializing the indices of the relation
-  for (int i = 0; i < proto.index_size(); i++) {
-    indices_.push_back(proto.index(i));
+  // Deserializing the index scheme defined for the relation, if any
```
Add the period symbol at the end of the comment to treat it as an English sentence.

saketj replied:
Done.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -93,9 +93,9 @@ CatalogRelation::CatalogRelation(const serialization::CatalogRelation &proto)
-    indices_.push_back(proto.index(i));
+  // Deserializing the index scheme defined for the relation, if any
+  if (proto.has_index_scheme()) {
+    setIndexScheme(IndexScheme::ReconstructFromProto(proto.index_scheme()));
```
I suggest to remove the method `setIndexScheme`, as it seems not to use by others.

saketj replied:
Done.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +243,70 @@ class CatalogRelation : public CatalogRelationSchema {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
-    indices_.emplace_back(name);
+  bool addIndex(const std::string &index_name, const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    // create an index_scheme, if it does not exist
```
Treat comments as sentences. Capitalize the first letter, and add the period symbol.

saketj replied:
Done.

zuyu commented:
@saketj Please review all your changes and treat comments as sentences by capitalizing the first letter and adding the period symbol.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +243,70 @@ class CatalogRelation : public CatalogRelationSchema {
+   * @return Whether the index exists or not.
    **/
-  bool hasIndexWithName(const std::string& name) {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
```
Why removed the mutex protection for readers?

saketj replied:
Fixed this.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +243,70 @@ class CatalogRelation : public CatalogRelationSchema {
+        SpinSharedMutexExclusiveLock<false> lock(layout_mutex_);
+        StorageBlockLayoutDescription *layout = default_layout_->getDescriptionMutable();
+        for (auto it = index_descriptions.begin(); it != index_descriptions.end(); ++it) {
+          // add a new index
```
Remove this over-documented comment.

saketj replied:
Done.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -310,18 +385,19 @@ class CatalogRelation : public CatalogRelationSchema {
   // The default layout for newly-created blocks.
   mutable std::unique_ptr<StorageBlockLayout> default_layout_;
+  // Mutex for locking the storage block layout.
+  mutable SpinSharedMutex<false> layout_mutex_;
```
Add the align info:

```
#include "storage/StorageConstants.hpp"

  alignas(kCacheLineBytes) mutable SpinSharedMutex<false> layout_mutex_;
```

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -310,18 +385,19 @@ class CatalogRelation : public CatalogRelationSchema {
+  // Defines a set of indices defined for this relation.
+  std::unique_ptr<IndexScheme> index_scheme_;
+  // Mutex for locking the index scheme.
+  mutable SpinSharedMutex<false> index_scheme_mutex;
```
Ditto for the align info.

saketj replied:
Align info added to the mutexes.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,98 @@
+#include "glog/logging.h"
+
+namespace quickstep {
+  bool IndexScheme::ProtoIsValid(const serialization::IndexScheme &proto) {
```
No indents for function names in the namespace:

```
namespace quickstep {

int foo() {
  int retval = 0;
  return retval;
}

}  // namespace quickstep
```

saketj replied:
Thanks. Have resolved this now.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+    * @param type The type of partitioning to be used to partition the
+    *             relation.
+    * @param num_partitions The number of partitions to be created.
+    * @param attribute The attribute on which the partitioning happens.
```
Remove comments for unused parameters.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+   *
+   * @param proto The Protocol Buffer serialization of a Index Scheme,
+   *              previously produced by getProto().
+   * @param attr_type The attribute type of the partitioning attribute.
```
Remove this comment.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+      for (std::size_t i = 0; i < index_descriptions.size(); ++i) {
+        // make a copy of the index_description before putting it in the map
+        IndexSubBlockDescription index_description(index_descriptions[i]);
+        value.emplace_back(index_description);
```
Try

```
  value.emplace_back(index_descriptions[i]);
```

as `emplace_back` calls the constructor.

saketj replied:
Good catch. Thank you for pointing this out. Have removed it now.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+
+ private:
+  // A map of index names to their index description
+  std::unordered_map<std::string, std::vector<IndexSubBlockDescription>> index_map_;
```
Do we need `mutex` protection for `index_map_`?

saketj replied:
I guess it should be find without one. Index scheme is only being accessed through CatalogRelation, which already protects the access to it through a mutex.

zuyu commented on parser/CMakeLists.txt:
```diff
@@ -91,6 +91,7 @@ add_library(quickstep_parser_ParseGeneratorTableReference ParseGeneratorTableRef
 add_library(quickstep_parser_ParseGroupBy ParseGroupBy.cpp ParseGroupBy.hpp)
 add_library(quickstep_parser_ParseHaving ParseHaving.cpp ParseHaving.hpp)
 add_library(quickstep_parser_ParseKeyValue ../empty_src.cpp ParseKeyValue.hpp)
+add_library(quickstep_parser_ParseIndexProperties ParseIndexProperties.cpp ParseIndexProperties.hpp)
```
Reorder this line before the above line.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.cpp:
```diff
@@ -0,0 +1,28 @@
+
+namespace quickstep {
+  // initialize constants for various index properties
+  const std::string BloomFilterIndexProperties::kBloomFilterSizeInBytes = "size";
```
Suggest to use `char *` instead for all strings in this file.

saketj replied:
@zuyu Is it because of the performance reason? I am using these constants to key into a string map and perform string comparisons on the stored string keys later. Shouldn't string make it easier for allocation/deallocation?

zuyu replied:
Yes, for the performance reasons.

After I checked the usages for these strings, and `valid_property_map_` and `hasInvalidCustomProperty` defined in the base class, there could have a better OO implementation here:
 1. mark `hasInvalidCustomProperty` virtual with the default true return value, and `BloomFilterIndexProperties` has its own implementation for size in bytes, num hashes, and project element count.
 1. remove `valid_property_map_` in the base class.

saketj replied:
@zuyu I see that you may be referring to the use of the `hasInvalidCustomProperty() `and `hasDuplicateCustomProperty()` by only the `BloomFilterIndexProperties` derived class. 
I intended them to be more of like helper routines that will be invoked by all the derived classes to check for duplicates and invalid properties. Otherwise this piece of code may be duplicated by all the other classes. 
Currently, CSBTreeIndex does not define much of additional tunable properties (at least that I know of),  but a BitWeavingIndex, or SMAIndex might define some. I thought this might come handy during that time. What do you suggest?

zuyu replied:
No, marking `hasInvalidCustomProperty` `virtual` should allow any derived classes to override the implementation.

But I think there could be a better way to check the unit for custom property.

Lastly, we should use c-like strings for static string members, instead of c++ `std::string`s.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+ */
+class IndexProperties {
+ public:
+  IndexProperties(bool is_index_description_valid,
```
Mark `is_index_description_valid` `const`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+ public:
+  IndexProperties(bool is_index_description_valid,
+                  const std::string &reason_for_invalid_description,
+                  const ParseKeyValue* invalid_property_node)
```
Change to `const ParseKeyValue *invalid_property_node` instead.

saketj replied:
done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    index_sub_block_description_(nullptr) {
+  }
+
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_type;
```
Move `typedef` before the constructor and after `public:`.

Read [Code Style Guide](https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide).

saketj replied:
Thanks. Have updated it now.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    return invalid_property_node_;
+  }
+
+  std::string getReasonForInvalidIndexDescription() const {
```
Return `const std::string&` instead to avoid deep copy of strings?

saketj replied:
Thanks, have updated it.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    return index_sub_block_description_.get();
+  }
+
+  void setIndexDescriptionAsInvalid(const std::string &reason, const ParseKeyValue *invalid_node = nullptr) {
```
Add the doxygen comments for all member functions in this file.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    invalid_property_node_ = invalid_node;
+  }
+
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
```
Why use a pointer instead of a reference?

saketj replied:
@zuyu This function is invoked from the ParseStatement, where a pointer is available to a key_value_list. (returned from the call to custom_properties_node->getKeyValueList()) 
I can de-reference the pointer and pass a reference here. Will it still give better performance?

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+  valid_property_map_type valid_property_map_;  // a map of index property constants and their associated type
+  bool is_index_description_valid_;
+  std::string reason_for_invalid_description_;
+  const ParseKeyValue* invalid_property_node_;  // referred object owned by ParseIndexProperties class
```
Change to `const ParseKeyValue *invalid_property_node_;`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+  : is_index_description_valid_(is_index_description_valid),
+    reason_for_invalid_description_(reason_for_invalid_description),
+    invalid_property_node_(invalid_property_node),
+    index_sub_block_description_(nullptr) {
```
Remove this line. It will use the default constructor.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+ protected:
+  valid_property_map_type valid_property_map_;  // a map of index property constants and their associated type
+  bool is_index_description_valid_;
+  std::string reason_for_invalid_description_;
```
Why use a string instead of some enum for this?

saketj replied:
The reason I didn't use enum was because this is supposed to be anyway a printable string and possible new indexes may define their own properties and corresponding validation and invalidation rules. A map would be additionally required if using an enum.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+    if (index_map_.find(index_name) != index_map_.end()) {
+      return false;  // index_name is already present!
+    } else {
+      std::string key = index_name;
```
Remove this line, and use `index_name` directly to avoid a deep copy of strings.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,193 @@
+    } else {
+      std::string key = index_name;
+      // value for this index_map key is the list of index descriptions provided
+      std::vector<IndexSubBlockDescription> value;
```
Use `index_subblock_descriptions` instead.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    // Define an existence map, which checks for duplicate properties in the key_value_list.
+    std::unordered_map<std::string, bool> is_duplicate_property;
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
```
Mark `key` `const`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+
+  bool hasInvalidCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
```
Mark `key` `const`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
```
Mark `key` `const`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+          setIndexDescriptionAsInvalid("size cannot be specified as a float", &key_value);
+          return false;
+        }
+        const int64_t bloom_filter_size = key_integer_value->value()->long_value();
```
Change to `std::int64_t`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+          setIndexDescriptionAsInvalid("num_hashes cannot be specified as a float", &key_value);
+          return false;
+        }
+        const int64_t number_of_hashes = key_integer_value->value()->long_value();
```
Change to `std::int64_t`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+          setIndexDescriptionAsInvalid("projected_element_count cannot be specified as a float", &key_value);
+          return false;
+        }
+        const int64_t projected_element_count = key_integer_value->value()->long_value();
```
Change to `std::int64_t`.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+class CSBTreeIndexProperties : public IndexProperties {
+ public:
+  CSBTreeIndexProperties()
+  : IndexProperties(true, "", nullptr) {
```
Four space indents.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,267 @@
+
+}  // namespace quickstep
+
+#endif  /* QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_ */
```
Change to

`#endif  // QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_`

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -195,22 +197,55 @@ class ParseStatementCreateIndex : public ParseStatement {
+     * @param index_properties_line_number
+     * @param index_properties_column_number
+     * @param index_properties
+     * TODO(ssaurabh) Fix comments
```
Fix the TODOs.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -195,22 +197,55 @@ class ParseStatementCreateIndex : public ParseStatement {
+                              const int index_properties_line_number,
+                              const int index_properties_column_number,
+                              PtrList<ParseKeyValue> *opt_index_properties)
+    : ParseStatement(line_number, column_number),
```
Four space indents.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -258,6 +293,26 @@ class ParseStatementCreateIndex : public ParseStatement {
+        return true;
+      } else {
+        return false;
+      }
```
Refactor to

```
  return opt_custom_properties_node_;
```

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -258,6 +293,26 @@ class ParseStatementCreateIndex : public ParseStatement {
+      return opt_custom_properties_node_.get();
+    }
+
+    const bool doesDefineCustomProperties() const {
```
Rename to `hasCustomProperties`.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
         container_child_fields->emplace_back();
-        for (const ParseString& attribute_name : *attribute_name_list_) {
-          container_child_fields->back().push_back(&attribute_name);
+        for (const ParseAttribute& attribute : *attribute_list_) {
```
Change to `const ParseAttribute &attribute`.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
 
-      if (attribute_name_list_.get() != nullptr) {
-        container_child_field_names->push_back("attribute_name_list");
+      if (attribute_list_.get() != nullptr) {
```
Remove `.get()`.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
       }
 
-      if (index_property_list_.get() != nullptr) {
+      if (opt_custom_properties_node_.get() != nullptr) {
```
Remove `.get()`.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
     std::unique_ptr<ParseString> index_type_;
-    std::unique_ptr<PtrList<ParseKeyValue> > index_property_list_;
+    std::unique_ptr<IndexProperties> index_properties_;
+    std::unique_ptr<ParseIndexProperties> opt_custom_properties_node_;
```
Rename to `custom_properties_node_`. You could add comments about it is optional.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
+          break;
+      }
+      // index_type must match one of the already defined indices
+      DCHECK(index_properties_.get() != nullptr);
```
Remove this line.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +331,44 @@ class ParseStatementCreateIndex : public ParseStatement {
+          index_properties_.reset(new CSBTreeIndexProperties());
+          break;
+        default:
+          break;
```
Add a `LOG(FATAL)` here.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,63 @@ void ExecutionGenerator::convertCreateIndex(
   // CreateIndex is converted to a CreateIndex operator.
   const CatalogRelationInfo *input_relation_info =
       findRelationInfoOutputByPhysical(physical_plan->input());
+  // NOTE: the object pointed to by input_relation is owned by CatalogRelation.
```
Remove this over-documented comment.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,63 @@ void ExecutionGenerator::convertCreateIndex(
+    const CatalogAttribute *catalog_attribute
+      = input_relation->getAttributeByName(attribute->attribute_name());
+    DCHECK(catalog_attribute != nullptr);
+    if (catalog_attribute != nullptr) {
```
This line is duplicated with the above check.

Remove this `if` statement.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -760,16 +760,63 @@ void ExecutionGenerator::convertCreateIndex(
+                                       catalog_attribute->getID());
+        break;
+      default:
+        break;
```
Change `break` to

```
  LOG(FATAL) << "Unknown index subblock type".
```

saketj replied:
Done.

zuyu commented on query_optimizer/physical/CreateIndex.hpp:
```diff
@@ -85,14 +101,18 @@ class CreateIndex : public Physical {
+   * @param input The input produces the relation to create index upon.
    * @param index_name The name of the index to create.
+   * @param index_attributes Set of attributes to create index upon.
+   * @param index_description A proto block describing the set of properties for this index
```
Add the period symbol at the end.

saketj replied:
Done.

zuyu commented on query_optimizer/physical/CreateIndex.hpp:
```diff
@@ -107,13 +127,20 @@ class CreateIndex : public Physical {
+        index_name_(index_name),
+        index_attributes_(index_attributes),
+        index_description_(index_description) {
         addChild(input_);
```
Reduce two space indents for `addChild`.

saketj replied:
Done.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -595,12 +595,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+
+  // Resolve attribute references
+  const PtrList<ParseAttribute> *index_attributes = create_index_statement.attribute_list();
+  const std::vector<E::AttributeReferencePtr> relation_attributes =
```
Use the reference instead to avoid deep copy of a vector.

saketj replied:
Done.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -595,12 +595,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+  // Resolve attribute references
+  const PtrList<ParseAttribute> *index_attributes = create_index_statement.attribute_list();
+  const std::vector<E::AttributeReferencePtr> relation_attributes =
+  input->getOutputAttributes();
```
Add four space indents.

saketj replied:
Done.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -595,12 +595,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+      bool is_resolved = false;
+      for (const E::AttributeReferencePtr &relation_attribute : relation_attributes) {
+        std::string relation_attr_name = relation_attribute->attribute_name();
+        std::string index_attr_name = index_attribute.attr_name()->value();
```
Mark above two strings as `const std::string &` to avoid deep copy of strings.

saketj replied:
Done.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -595,12 +595,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+    std::unique_ptr<IndexSubBlockDescription> index_description(new IndexSubBlockDescription());
+    index_description->CopyFrom(*create_index_statement.getIndexProperties()->getIndexDescription());
+    index_description_shared.reset(index_description.release());
+    DCHECK(index_description_shared.get() != nullptr);
```
Remove `.get()`.

saketj replied:
Done.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -595,12 +595,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+    // create a deep copy of the index description and pass its ownership to the shared ptr
+    std::unique_ptr<IndexSubBlockDescription> index_description(new IndexSubBlockDescription());
+    index_description->CopyFrom(*create_index_statement.getIndexProperties()->getIndexDescription());
+    index_description_shared.reset(index_description.release());
```
Why not directly reset `index_description_shared`, instead of indirectly using `unique_ptr`?

saketj replied:
@zuyu I am using the unique_ptr to avoid using the raw pointers. An alternative could have been `IndexSubBlockDescription *index_description = new IndexSubBlockDescription()
index_description_shared.reset(index_description)`
Thought that using raw pointers in this way, would be a bad practice. Also, I need to create a heap object because the object will be then shared by physical plan too.

zuyu replied:
Please ignore this.

I missed the const quantifier in the shared pointer.

saketj commented:
@zuyu Thanks Zuyu for the detailed review and comments. These were very helpful and I have fixed the issues raised in the comments. This has served as a very good exercise for me to pay attention to minor performance issues that can be introduced by even simply missing on a reference usage. Thanks again.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,186 @@
+   /**
+    * @brief Constructor.
+    *
+    **/
```
Reduce one space indent for the comments above, and remove the middle line 52.

saketj replied:
Fixed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+   * @brief Constructor.
+   **/
+  BloomFilterIndexProperties()
+    : IndexProperties(false, "Bloom Filter index is not yet implemented", nullptr) {
```
Two additional space indents here.

saketj replied:
Fixed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+  * @brief Constructor.
+  **/
+  CSBTreeIndexProperties()
+    : IndexProperties(true, "", nullptr) {
```
Two more space indents here.

saketj replied:
Fixed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+  ParseIndexProperties(const int line_number,
+                       const int column_number,
+                       PtrList<ParseKeyValue> *key_value_list)
+    : ParseTreeNode(line_number, column_number),
```
Two more space indents.

saketj replied:
Fixed.

zuyu commented:
@saketj You may want to reconsider the alternative implementation for `hasInvalidCustomProperty` in `parser/ParseIndexProperties.hpp`.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+class IndexProperties {
+ public:
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_type;
+  /**
```
Add a new line between `typedef` and the comment.

saketj replied:
Fixed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+  IndexProperties(const bool is_index_description_valid,
+                  const std::string &reason_for_invalid_description,
+                  const ParseKeyValue* invalid_property_node = nullptr)
+    : is_index_description_valid_(is_index_description_valid),
```
Add two more space indents here.

saketj replied:
Fixed

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+   *
+   * @return True if index description is valid, false otherwise.
+   **/
+  bool hasValidIndexDescription() const {
```
Rename to `isValidIndexDescription`?

saketj replied:
Fixed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,329 @@
+  static const std::string kBloomFilterSizeInBytes;    // is of type integer
+  static const std::string kBloomFilterNumHashes;      // is of type integer
+  static const std::string kBloomFilterProjectElementCount;  // is of type integer
+  /**
```
Add a new line between the static member and the comment.

saketj replied:
Fixed.

saketj commented:
Multiple documentation fixes squashed into the previous commit.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -170,6 +172,45 @@ class CatalogRelation : public CatalogRelationSchema {
+   * @param index_scheme The index scheme object for a relation, which
+   *        becomes owned by this relation.
+   **/
+  void setIndexScheme(IndexScheme* index_scheme);
```
Unused method?

saketj replied:
Removed.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,98 @@
+      index_descriptions.emplace_back(index_entry.index_description(i));
+    }
+    // store the index_name and corresponding list of index descriptions in map
+    index_scheme->index_map_[index_name] = index_descriptions;
```
Remove the above `index_name`definition, and use `index_entry.index_name()`, to avoid deep copy of strings.

Refactor this line to `move` `index_descriptions` instead of making a copy.

saketj replied:
Refactored.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,98 @@
+
+    // populate the details of the index entry
+    index_entry->set_index_name(cit->first);
+    std::vector<IndexSubBlockDescription> index_descriptions = cit->second;
```
Mark it `const reference` instead of making a copy.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,98 @@
+    // populate the details of the index entry
+    index_entry->set_index_name(cit->first);
+    std::vector<IndexSubBlockDescription> index_descriptions = cit->second;
+    for (auto index_description_itr = index_descriptions.begin();
```
Rename `index_description_itr` to `it`?

saketj replied:
Renamed.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,185 @@
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map.
+    for (auto cit = index_map_.cbegin(); cit != index_map_.cend(); ++cit) {
+      std::vector<IndexSubBlockDescription> index_descriptions = cit->second;
```
Use `const reference` instead.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,185 @@
+           ++index_description_itr) {
+        const IndexSubBlockDescription &stored_description = *index_description_itr;
+        // Check if the stored description matches as the given description.
+        if (areIndexDescriptionsSimilar(stored_description, index_description)) {
```
Use `*index_description_itr` directly, and remove above definition for `stored_description`.

Rename `index_description_itr` to `description_it`?

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,185 @@
+                        const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    if (index_map_.find(index_name) != index_map_.end()) {
+      return false;  // index_name is already present!
+    } else {
```
Remove the `else` line, and reduce two space indents for codes in the `else` part.

saketj replied:
Removed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,331 @@
+   * @param reason A printable string that describes the reason for invalidity.
+   * @param invalid_node A parse node that may have caused this index to become invalid.
+   **/
+  void setIndexDescriptionAsInvalid(const std::string &reason, const ParseKeyValue *invalid_node = nullptr) {
```
I prefer `enums` to indicate error codes, instead of `string`s.

saketj replied:
Done.

saketj commented:
@zuyu I have refactored the ParseIndexProperties now to use enums instead of strings and replaced the string constants with const char*. I also removed some of the methods to make the OO abstraction more cleaner now.

zuyu commented:
@saketj You need to pass the CI first.

saketj commented:
@zuyu Seems like only one of the Travis build jobs failed due to a non-zero ref count for newly added ExecutionGenerator test case. Not sure if the commits to this PR have introduced this. I have restarted the Travis Build, to see if it fixes this.

zuyu commented:
@saketj No, the known, existing block ref count issue is not related to this PR.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +233,70 @@ class CatalogRelation : public CatalogRelationSchema {
-      return false;
-    }
+  bool hasIndexWithName(const std::string &index_name) const {
+    SpinSharedMutexExclusiveLock<false> lock(index_scheme_mutex);
```
Change to `SpinSharedMutexSharedLock` for readers.

FYI, `SpinSharedMutexExclusiveLock` is used for writers.

saketj replied:
Fixed.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +233,70 @@ class CatalogRelation : public CatalogRelationSchema {
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    SpinSharedMutexExclusiveLock<false> lock(index_scheme_mutex);
```
Use `SpinSharedMutexSharedLock` instead.

saketj replied:
Fixed.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +233,70 @@ class CatalogRelation : public CatalogRelationSchema {
+   * @return Whether the index was added successfully or not
    **/
-  void addIndex(const std::string& name) {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
```
Add back this writer lock.

zuyu replied:
Add back this writer lock.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +233,70 @@ class CatalogRelation : public CatalogRelationSchema {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
-    indices_.emplace_back(name);
+  bool addIndex(const std::string &index_name, const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    // Create an index_scheme, if it does not exist.
```
Add the reader lock `SpinSharedMutexSharedLock`.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -203,35 +233,70 @@ class CatalogRelation : public CatalogRelationSchema {
+
+    // Acquire the lock and add the index.
+    {
+      SpinSharedMutexExclusiveLock<false> lock(index_scheme_mutex);
```
Once adding the writer lock at the beginning, remove this lock.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,97 @@
+
+  // Check that each index entry contains initialized valid index descriptions
+  for (int i = 0; i < proto.index_entries_size(); ++i) {
+    serialization::IndexScheme_IndexEntry index_entry = proto.index_entries(i);
```
Change to `const serialization::IndexScheme::IndexEntry &index_entry` to avoid a deep copy.

saketj replied:
Done. The unique index_name is being checked by the ReconstructFromProto() function now, which creates the index_map object.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,97 @@
+    serialization::IndexScheme_IndexEntry index_entry = proto.index_entries(i);
+    for (int j = 0; j < index_entry.index_description_size(); ++j) {
+      IndexSubBlockDescription index_description = index_entry.index_description(j);
+      if (!index_description.IsInitialized()) {
```
This check is not sufficient.

What we want to see is that

```
for (int j = 0; j < index_entry.index_description_size(); ++j) {
  if (!IndexSubBlockDescription::ProtoIsValid(index_entry.index_description(j), relation)) {
    return false;
  }
}
```

In the static method `IndexSubBlockDescription::ProtoIsValid`, you need to check whether each `indexed_attribute_id` is valid in `relation`. You may want to add this new method in a new file under `/storage` called `IndexSubBlockDescription.[hc]pp`.

saketj replied:
@zuyu Using the CatalogRelation object itself may not be appropriate at this point, because this function is invoked from the CatalogRelation constructor itself. However, I have now implemented a more stricter version of this check in the CatalogRelation constructor itself. Each index description defined in the index_scheme is now being checked for its presence in the index descriptions defined in the StorageBlockLayout.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,97 @@
+  std::unique_ptr<IndexScheme> index_scheme(new IndexScheme());
+
+  for (int index_num = 0; index_num < proto.index_entries_size(); ++index_num) {
+    serialization::IndexScheme_IndexEntry index_entry = proto.index_entries(index_num);
```
Use the const reference to avoid deep copy.

saketj replied:
Fixed.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,97 @@
+      index_descriptions.emplace_back(index_entry.index_description(i));
+    }
+    // store the index_name and corresponding list of index descriptions in map
+    index_scheme->index_map_[index_entry.index_name()] = std::move(index_descriptions);
```
Refactor to `index_scheme->index_map_.emplace(index_entry.index_name(), std::move(index_descriptions))` to avoid calling the default constructor for the key field.

saketj replied:
Fixed. Added check for uniqueness. Thanks.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  /**
+   * @brief Virtual destructor.
```
It's not `virtual`.

saketj replied:
Removed.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+   * @param desc_b Second index description.
+   * @return Whether the two index_descriptions are similar or not.
+   **/
+  bool areIndexDescriptionsSimilar(const IndexSubBlockDescription &desc_a,
```
Rename to `areSameIndexDescriptions`? I think `similar` is ambiguous.

Also, rename the parameters to `description_expected` and `description_checked`, respectively. The former is for the known description, and the latter for the one used for comparison.

saketj replied:
Refactored.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+    if (desc_a.sub_block_type() != desc_b.sub_block_type()) {
+      return false;
+    }
+    // TODO(@ssaurabh): add condition to check for matching attribute ids for both indices
```
Please implement the TODO.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+   **/
+  bool addIndexMapEntry(const std::string &index_name,
+                        const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    if (index_map_.find(index_name) != index_map_.end()) {
```
Do you need any protections for writers?

saketj replied:
This code is protected by lock from within the CatalogRelation addIndex() function.

zuyu replied:
Then add comments about `@note The caller must hold locks before calling this method.`

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,392 @@
+   *                              to become valid.
+   **/
+  IndexProperties() {
+  }
```
Need to add the `virtual destructor`.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,392 @@
+    valid_property_map_[kBloomFilterNumHashes] = ParseKeyValue::KeyValueType::kStringInteger;
+    valid_property_map_[kBloomFilterProjectElementCount] = ParseKeyValue::KeyValueType::kStringInteger;
+  }
+
```
Need to add overridden destructor.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,392 @@
+    index_sub_block_description_.reset(new IndexSubBlockDescription());
+    index_sub_block_description_->set_sub_block_type(IndexSubBlockDescription::CSB_TREE);
+  }
+
```
Need to add the overridden destructor.

saketj replied:
Added.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,209 @@
+   * @return Whether the two index_descriptions are similar or not.
+   **/
+  bool areIndexDescriptionsSame(const IndexSubBlockDescription &description_expected,
+                                   const IndexSubBlockDescription &description_checked) const {
```
Align with `(`.

saketj replied:
Done.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    // Determine the type of attribute_extension based on the index type.
+    auto *attribute_extension = &(CSBTreeIndexSubBlockDescription::indexed_attribute_id);  // default
```
Please rewrite this function after #94 gets merged.

saketj replied:
Accounted for changes to StorageBlockLayout

zuyu commented on storage/StorageBlockLayout.proto:
```diff
@@ -83,6 +84,22 @@ message SMAIndexSubBlockDescription {
+    // - Default number of elements expected to be stored = 1 million.
+    optional fixed64 bloom_filter_size = 34 [default = 10000];
+    optional int32 number_of_hashes = 35 [default = 20];
+    optional fixed64 projected_element_count = 36 [default = 1000000];
```
These numbers (from 33 to 36) assigned to the fields does not follow the rules described in `IndexSubBlockDescription`.

saketj replied:
Fixed. Numbers updated to start from 96.

zuyu commented on storage/StorageBlockLayout.proto:
```diff
@@ -73,6 +74,21 @@ message IndexSubBlockDescription {
   extensions 32 to max;
 }
 
+
```
Remove this empty line.

saketj replied:
Done.

zuyu commented on storage/StorageBlockLayout.proto:
```diff
@@ -73,6 +74,21 @@ message IndexSubBlockDescription {
+    optional fixed64 projected_element_count = 98 [default = 1000000];
+  }
+}
+
```
Add one more empty line here to separate two different message types.

saketj replied:
Added.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -810,13 +810,45 @@ void ExecutionGenerator::convertCreateIndex(
+      // Check if the given index description is valid.
+      THROW_SQL_ERROR() << "The index with given properties cannot be created.";
+    } else {
+      index_descriptions.emplace_back(index_description);
```
Change to the following to take advantages of the `move` semantics:

```
  index_descriptions.push_back(move(index_description));
```

saketj replied:
updated.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -131,6 +140,7 @@ void CatalogRelation::setPartitionScheme(PartitionScheme* partition_scheme) {
   partition_scheme_.reset(partition_scheme);
 }
 
+
```
Remove this empty line.

saketj replied:
Removed.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -145,8 +155,8 @@ serialization::CatalogRelation CatalogRelation::getProto() const {
-  for (std::vector<std::string>::const_iterator it = indices_.begin(); it != indices_.end(); ++it) {
-    proto.add_index(*it);
+  if (hasIndexScheme()) {
+    proto.mutable_index_scheme()->CopyFrom(index_scheme_->getProto());
```
It is ok to use `MergeFrom` instead of `CopyFrom`.

saketj replied:
Replaced.

zuyu replied:
I think you replaced another place, but it is up to you to change it as well.

saketj replied:
Replaced.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,101 @@
+  for (int i = 0; i < proto.index_entries_size(); ++i) {
+    const serialization::IndexScheme_IndexEntry &index_entry = proto.index_entries(i);
+    for (int j = 0; j < index_entry.index_description_size(); ++j) {
+      IndexSubBlockDescription index_description = index_entry.index_description(j);
```
Use the reference here to avoid a deep copy:

```
  const IndexSubBlockDescription &index_description = index_entry.index_description(j);
```

saketj replied:
Updated.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,101 @@
+    const serialization::IndexScheme_IndexEntry &index_entry = proto.index_entries(i);
+    for (int j = 0; j < index_entry.index_description_size(); ++j) {
+      IndexSubBlockDescription index_description = index_entry.index_description(j);
+      if (!index_description.IsInitialized()) {
```
This check is not enough for various implementations of `IndexSubBlockDescription`, including `CS-b-tree`, `SMA`, and `bloom filter`. Instead, you may want to add the static method called `ProtoIsValid` in a new file called `IndexSubBlockDescriptionFactory` under `/storage`.

saketj replied:
@zuyu As I was adding this IndexSubBlockDescriptionFactory, I realized that there is a method  `LayoutDescriptionIsValid()` in `SubBlockTypeRegistry.cpp` that precisely checks the integrity for each index type, among other things. The implementation is provided by each index type eg. `CSBTreeIndexSubBlock::DescriptionIsValid()`. Adding IndexSubBlockDescriptionFactory will duplicate this check. 
So instead, I have now refactored to just verify that StorageBlockLayout and IndexScheme contain the exact same IndexSubBlockDescriptions (through protobuf serialized strings comparison for exact match). The integrity check of IndexSubBlockDescription is a part of StorageBlockLayoutDescription validity check now.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,195 @@
+   * @brief Destructor.
+   **/
+  ~IndexScheme() {
+  }
```
It is up to you to keep or remove the default constructor and destructor, provided by the compiler by default.

saketj replied:
Removed in favor of default.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,195 @@
+   *
+   * @return The number of indices defined for the relation.
+   **/
+  inline const std::size_t getNumIndices() const {
```
Remove the first `const`.

saketj replied:
removed.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,195 @@
+    // Check if there are matching attribute ids for both indices of same type.
+    // Create a hash map of attribute ids in description_expected,
+    // and probe the hash map for duplicates using the attribute ids from description_checked.
+    std::unordered_map<std::int32_t, bool> existence_map;
```
Why need the second `bool` part here? I think an `unordered_set` of `attribute_id`s is good enough to find the duplicates from `description_checked`.

saketj replied:
Replaced with serialized protobuf message string comparison for more robust check.

zuyu commented on parser/ParseIndexProperties.cpp:
```diff
@@ -0,0 +1,29 @@
+
+namespace quickstep {
+  // Initialize constants for various index properties.
+
```
Remove this empty line.

saketj replied:
Removed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+    valid_property_map_[kBloomFilterProjectElementCount] = ParseKeyValue::KeyValueType::kStringInteger;
+  }
+
+  ~BloomFilterIndexProperties() {
```
Add `override` between `)` and `{`.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+    index_sub_block_description_->set_sub_block_type(IndexSubBlockDescription::CSB_TREE);
+  }
+
+  ~CSBTreeIndexProperties() {
```
Add `override` between `)` and `{`.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+    index_sub_block_description_.reset();
+  }
+};
+
```
Add an empty line to separate two classes.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+ protected:
+  std::unique_ptr<IndexSubBlockDescription> index_sub_block_description_;
+};
+
```
Add one more empty line to separate two classes.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+    return invalid_property_node_;
+  }
+
+  const std::string getReasonForInvalidIndexDescription() const override {
```
Remove the first `const`.

saketj replied:
removed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,404 @@
+  InvalidIndexType invalid_index_type_;
+  const ParseKeyValue *invalid_property_node_;  // referred object owned by ParseIndexProperties class.
+};
+
```
Add one more empty line to separate two classes.

saketj replied:
Added.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -245,8 +279,8 @@ class ParseStatementCreateIndex : public ParseStatement {
      **/
-    const PtrList<ParseString>& attribute_name_list() const {
-      return *attribute_name_list_;
+    const PtrList<ParseAttribute>* attribute_list() const {
```
Why change the return value from a reference to a pointer? Could `attribute_list_` be `nullptr`?

saketj replied:
@zuyu  yeah, the nullptr was the reason to change it to a pointer.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -258,6 +292,22 @@ class ParseStatementCreateIndex : public ParseStatement {
+    }
+
+    const ParseIndexProperties* getCustomPropertiesNode() const {
+      return custom_properties_node_.get();
```
For above two functions, prefer to return the reference, unless they could be `NULL`.

saketj replied:
These can be null.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -258,6 +292,22 @@ class ParseStatementCreateIndex : public ParseStatement {
+      return custom_properties_node_.get();
+    }
+
+    const bool hasCustomProperties() const {
```
Remove the first `const`.

saketj replied:
Removed.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +326,44 @@ class ParseStatementCreateIndex : public ParseStatement {
     std::unique_ptr<ParseString> index_name_;
     std::unique_ptr<ParseString> relation_name_;
-    std::unique_ptr<PtrList<ParseString> > attribute_name_list_;
+    std::unique_ptr<PtrList<ParseAttribute> > attribute_list_;
```
Remove the space between two `>`s, as we are using `c++11`.

saketj replied:
Space removed.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +326,44 @@ class ParseStatementCreateIndex : public ParseStatement {
+    std::unique_ptr<ParseIndexProperties> custom_properties_node_;
+
+    void initializeIndexType() {
+      int index_type_enum_val = std::stoi(index_type_->value());
```
Mark it `const`.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +326,44 @@ class ParseStatementCreateIndex : public ParseStatement {
+          index_properties_.reset(new CSBTreeIndexProperties());
+          break;
+        default:
+          LOG(FATAL) << "Unknown index subblock type.";
```
Don't we need to support `SMAIndex` as well?

saketj replied:
@zuyu As per previous discussions, it came out that SMAIndexes will be created as system-default and will not required to be created through a separate create index statement, since they have negligible cost to build. I can add that, if it seems likely to be also created through CREATE INDEX. Will also have to add IndexProperties class corresponding to it too, in that case.

zuyu replied:
Then you may want to add `SMA` case with a different `glog` message for the known index subblock type.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -116,13 +111,26 @@ CatalogRelation::CatalogRelation(const serialization::CatalogRelation &proto)
+    // Ensure that indices present in the block layout are the same as in the index scheme.
+    const std::size_t num_indices_expected = index_scheme_->getNumIndices();
+    const std::size_t num_indices_checked = proto.default_layout().index_description_size();
+    DCHECK(num_indices_expected == num_indices_checked);
```
Change to `DCHECK_EQ`.

saketj replied:
Fixed.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -116,13 +111,26 @@ CatalogRelation::CatalogRelation(const serialization::CatalogRelation &proto)
+    const std::size_t num_indices_expected = index_scheme_->getNumIndices();
+    const std::size_t num_indices_checked = proto.default_layout().index_description_size();
+    DCHECK(num_indices_expected == num_indices_checked);
+    for (int i = 0; i < proto.default_layout().index_description_size(); ++i) {
```
We could use `num_indices_checked`, instead of `proto.default_layout().index_description_size()`.

saketj replied:
Refactored.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -316,12 +376,11 @@ class CatalogRelation : public CatalogRelationSchema {
+  // Defines a set of indices defined for this relation.
+  std::unique_ptr<IndexScheme> index_scheme_;
+  // Mutex for locking the index scheme.
+  alignas(kCacheLineBytes) mutable SpinSharedMutex<false> index_scheme_mutex;
```
Rename to `index_scheme_mutex_` for the class private member.

saketj replied:
Refactored.

zuyu commented on catalog/IndexScheme.cpp:
```diff
@@ -0,0 +1,101 @@
+    const serialization::IndexScheme_IndexEntry &index_entry = proto.index_entries(i);
+    for (int j = 0; j < index_entry.index_description_size(); ++j) {
+      const IndexSubBlockDescription &index_description = index_entry.index_description(j);
+      if (!index_description.IsInitialized()) {
```
I believe I have mentioned before that such check for `IndexSubBlockDescription` is not sufficient enough for three implementations. Instead, we need to add a new static method called `ProtoIsValid` (maybe in a new file `/storage/IndexSubBlockFactory` to avoid cycle dependency).

Have you added that? Thanks!

saketj replied:
Added a separate Factory implementation to check for index description validity.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+  // A map of index names to their index description.
+  std::unordered_map<std::string, std::vector<IndexSubBlockDescription>> index_map_;
+
+ private:
```
Remove this duplicate `private`.

saketj replied:
Removed.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,183 @@
+ private:
+  DISALLOW_COPY_AND_ASSIGN(IndexScheme);
+};
+/** @} */
```
Add one empty line between line 178 and 179.

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+   * @return Returns true if the given set of custom properties was valid
+   *         and was applied correctly, otherwise returns false.
+   **/
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
```
Move this virtual function before `getIndexDescription`.

saketj replied:
Updated.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+ public:
+  // An enum specifying the reason for an invalid index description, if any.
+  enum class InvalidIndexType {
+    kNone,
```
Change to `kNone = 0,`.

zuyu replied:
Actually, I think it's better to move this enum class in the base class, as you have defined it twice for two derived classes.

saketj replied:
Refactored and moved enums in the base class.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+
+#include <cstdint>
+#include <memory>
+#include <set>
```
Remove the unused `<set>` header, and run `iwyu` for all new files. Thanks!

saketj replied:
Updated.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+  valid_property_map_type valid_property_map_;  // a map of index property constants and their associated type.
+  const ParseKeyValue *invalid_property_node_;  // referred object owned by ParseIndexProperties class.
+
+  void setIndexDescriptionAsInvalid(InvalidIndexType invalid_index_type,
```
Mark it `const`.

zuyu replied:
As the follow-up to move the enum class for the invalid index type to the base class, it is also advisable to move this method as a `protected` one in the base class.

saketj replied:
Refactored and moved to base class.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+ public:
+  // An enum specifying the reason for an invalid index description, if any.
+  enum class InvalidIndexType {
+    kNone,
```
Change to `kNone = 0,`.

zuyu replied:
Suggest to move it to the base class, including `invalid_index_type_`.

saketj replied:
Updated.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,407 @@
+
+ protected:
+  std::unique_ptr<IndexSubBlockDescription> index_sub_block_description_;
+};
```
Do you intend to allow this class and its derived classes copyable? If not, add `DISALLOW_COPY_AND_ASSIGN(IndexProperties)` and do so for all derived classes.

saketj replied:
Added disallow for copy and assign.

zuyu commented:
@saketj I have some OO-design suggestions for `IndexProperties` and its derived classes in `/parser/ParseIndexProperties.hpp`.

 1. move the enum classes for the invalid property types defined in two derived classes to the base class.
 1. move `invalid_index_type_` and `invalid_property_node_` to the base class as `protected` members.
 1. move `setIndexDescriptionAsInvalid` to the base class as a `protected` method.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -810,13 +810,45 @@ void ExecutionGenerator::convertCreateIndex(
-  } else {
-    execution_plan_->addRelationalOperator(
-       new CreateIndexOperator(input_relation, physical_plan->index_name()));
+    << " already has an index named "<< physical_plan->index_name();
```
Add four whitespace indents here.

saketj replied:
Added.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -810,13 +810,45 @@ void ExecutionGenerator::convertCreateIndex(
+  }
+
+  // Assertion: must have at least one attribute specified to build index upon.
+  DCHECK_GT(physical_plan->index_attributes().size(), 0u);
```
Remove the over-documented comment above for the check.

saketj replied:
Removed.

zuyu commented on relational_operators/CreateIndexOperator.cpp:
```diff
@@ -28,11 +28,15 @@ bool CreateIndexOperator::getAllWorkOrders(WorkOrdersContainer *container,
+  if (!work_generated_) {
+    work_generated_ = true;
+    relation_->addIndex(index_name_, index_descriptions_);
+  }
```
Bug! Remove all code in the `if` statement, as it is duplicated in 
`CreateIndexOperator::updateCatalogOnCompletion` below.

Don't you have a test to cover that one relation is added the same index twice?

Finally, you may want to mark the `catalog` with a new `CatalogDatabase::Status` called `pendingIndexAdditions` (defined in `/catalog/CatalogDatabase.hpp`) using `CatalogDatabase::setStatus`.

saketj replied:
Updated and added test cases.

zuyu commented on query_optimizer/logical/CreateIndex.cpp:
```diff
@@ -36,11 +36,17 @@ void CreateIndex::getFieldStringItems(
+  container_child_fields->push_back(CastSharedPtrVector<OptimizerTreeBase>(index_attributes_));
+
+  inline_field_names->push_back("serialized_index_description");
+  inline_field_values->push_back(index_description_->SerializeAsString());
```
In terms of a human readable form, I think `DebugString` or `ShortDebugString` is better than `SerializeAsString`.

saketj replied:
Replaced with DebugString.

zuyu commented:
@saketj You need to add test cases for creating indices in at least two places: `/catalog/tests/Catalog_unittest.cpp` and `/query_optimizer/tests/execution_generator/Create.test`.

zuyu commented on query_optimizer/logical/CreateIndex.hpp:
```diff
@@ -59,18 +61,32 @@ class CreateIndex : public Logical {
+  /**
+   * @return Shared pointer to the index properties.
+   */
+  const std::shared_ptr<const IndexSubBlockDescription> index_description() const {
```
Remove the first `const`, and I believe you borrowed the implementation for `CreateTable`, but I think we don't need the first one.

saketj replied:
Updated.

zuyu commented on query_optimizer/logical/CreateIndex.hpp:
```diff
@@ -99,13 +119,20 @@ class CreateIndex : public Logical {
+        index_name_(index_name),
+        index_attributes_(index_attributes),
+        index_description_(index_description) {
         addChild(input_);
```
Reduce four whitespace indents.

saketj replied:
Removed.

zuyu commented on query_optimizer/logical/CreateIndex.hpp:
```diff
@@ -99,13 +119,20 @@ class CreateIndex : public Logical {
   }
 
   LogicalPtr input_;
   std::string index_name_;
```
Why not mark as `const`?

saketj replied:
Updated.

zuyu commented on query_optimizer/logical/CreateIndex.hpp:
```diff
@@ -99,13 +119,20 @@ class CreateIndex : public Logical {
 
   LogicalPtr input_;
   std::string index_name_;
+  std::vector<expressions::AttributeReferencePtr> index_attributes_;
```
Why not mark as `const`?

saketj replied:
Updated.

zuyu commented on query_optimizer/physical/CreateIndex.cpp:
```diff
@@ -35,11 +35,17 @@ void CreateIndex::getFieldStringItems(
+  container_child_fields->push_back(CastSharedPtrVector<OptimizerTreeBase>(index_attributes_));
+
+  inline_field_names->push_back("serialized_index_description");
+  inline_field_values->push_back(index_description_->SerializeAsString());
```
As mentioned above, `DebugString` or `ShortDebugString` could be better.

saketj replied:
Replaced with DebugString.

zuyu commented on query_optimizer/physical/CreateIndex.hpp:
```diff
@@ -61,18 +63,32 @@ class CreateIndex : public Physical {
+  /**
+   * @return Shared pointer to the index properties.
+   */
+  const std::shared_ptr<const IndexSubBlockDescription> index_description() const {
```
Remove the first `const`.

saketj replied:
removed.

zuyu commented on query_optimizer/physical/CreateIndex.hpp:
```diff
@@ -107,13 +127,20 @@ class CreateIndex : public Physical {
+        index_name_(index_name),
+        index_attributes_(index_attributes),
+        index_description_(index_description) {
+      addChild(input_);
```
Reduce two whitespace indents.

saketj replied:
Updated.

zuyu commented on query_optimizer/physical/CreateIndex.hpp:
```diff
@@ -107,13 +127,20 @@ class CreateIndex : public Physical {
 
   PhysicalPtr input_;
   std::string index_name_;
+  std::vector<expressions::AttributeReferencePtr> index_attributes_;
```
Mark above two as `const`.

saketj replied:
Updated.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -596,12 +596,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+      }
+      if (!is_resolved) {
+        THROW_SQL_ERROR_AT(&index_attribute)<< "Attribute "<< index_attribute.attr_name()->value()
+        << " is undefined for the relation "<< create_index_statement.relation_name()->value();
```
Add four whitespace indents.

saketj replied:
Updated.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -596,12 +596,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+        }
+      }
+      if (!is_resolved) {
+        THROW_SQL_ERROR_AT(&index_attribute)<< "Attribute "<< index_attribute.attr_name()->value()
```
Add whitespace before `<<`.

saketj replied:
Added.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -596,12 +596,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+    if (create_index_statement.getIndexProperties()->getInvalidPropertyNode() != nullptr) {
+      // if exact location is known in the parser, the error is thrown at that specific node
+      THROW_SQL_ERROR_AT(create_index_statement.getIndexProperties()->getInvalidPropertyNode())
+        << create_index_statement.getIndexProperties()->getReasonForInvalidIndexDescription();
```
Add two more whitespace indents.

saketj replied:
Updated.

zuyu commented on query_optimizer/resolver/Resolver.cpp:
```diff
@@ -596,12 +596,63 @@ StorageBlockLayoutDescription* Resolver::resolveBlockProperties(
+    } else {
+      // else the error is thrown at the index name node
+      THROW_SQL_ERROR_AT(create_index_statement.index_type())
+        << create_index_statement.getIndexProperties()->getReasonForInvalidIndexDescription();
```
FYI, you could simplify the whole `if-else` code fragment by keeping a reference to `create_index_statement.getIndexProperties()`.

saketj replied:
Refactored.

zuyu commented on relational_operators/CreateIndexOperator.hpp:
```diff
@@ -75,6 +84,8 @@ class CreateIndexOperator : public RelationalOperator {
   CatalogRelation *relation_;
   const std::string &index_name_;
+  const std::vector<IndexSubBlockDescription> index_descriptions_;
+  bool work_generated_;
```
Remove this unnecessary member.

saketj replied:
Removed.

zuyu commented on relational_operators/CreateIndexOperator.hpp:
```diff
@@ -75,6 +84,8 @@ class CreateIndexOperator : public RelationalOperator {
  private:
   CatalogRelation *relation_;
   const std::string &index_name_;
+  const std::vector<IndexSubBlockDescription> index_descriptions_;
```
We could use the reference instead.

saketj replied:
I tried using a reference here. However, sometimes the index_descriptions_ vector object, when invoked in a different thread from CreateIndexOperator, is cleared out. I am investigating this and meanwhile have left this unchanged.

zuyu replied:
I see, then we may want to use the rvalue via `std::move` in the constructor to avoid a deep copy.

saketj commented:
@zuyu  Thanks for the comments. I have fixed the changes from the review and updated the tests for the CREATE INDEX now.

zuyu commented on query_optimizer/tests/execution_generator/Create.test:
```diff
@@ -38,3 +38,24 @@ CREATE TABLE foo2 (col1 INT, col2 VARCHAR(80))
+--
+ERROR: The relation foo3 already defines this index on col1
+==
+CREATE INDEX idx2 ON foo3(col2) USING CSBTREE;
```
We need to add tests for `bloom-filter` index as well.

saketj replied:
Added.

zuyu commented:
@saketj It looks good, except for that we need to add the same tests for adding indices in the following places as well:
 1. `/parser/tests/Create.test`
 1. `/query_optimizer/tests/resolver/Create.test`
 1. `/query_optimizer/tests/logical_generator/Create.test`
 1. `/query_optimizer/tests/physical_generator/Create.test`

saketj commented:
@zuyu I have added the additional tests for CREATE INDEX now at the following places (and moved the original ones from Create.test files to their individual Index.test files)

1. /parser/tests/Index.test
2. /query_optimizer/tests/resolver/Index.test
3. /query_optimizer/tests/logical_generator/Index.test
4. /query_optimizer/tests/physical_generator/Index.test
5. /query_optimizer/tests/execution_generator/Index.test

Also, I have squashed some of the git commits now to make the history for this PR look cleaner.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,177 @@
+  /**
+   * @brief Adds a new index entry to the index map.
+   * @warning Must call before hasIndexWithName() and hasIndexWithDescription().
+   *
```
Add the following comment here:

```
   * @note This method assumes that the caller has already acquired the
   *       necessary locks before invoking it.
```

saketj replied:
Added.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,177 @@
+      // Make a copy of the index_description before putting it in the map.
+      index_sub_block_descriptions.emplace_back(index_descriptions[i]);
+    }
+    index_map_[index_name] = index_sub_block_descriptions;
```
Use `emplace` instead to avoid the default value and then a deep copy.

saketj replied:
Updated.

zuyu commented on catalog/tests/Catalog_unittest.cpp:
```diff
@@ -506,4 +507,40 @@ TEST_F(CatalogTest, CatalogSplitRowStoreSerializationTest) {
+  EXPECT_TRUE(rel->addIndex("idx2", index_descriptions));
+  EXPECT_TRUE(rel->hasIndexWithName("idx2"));
+
+  checkCatalogSerialization();
```
In `CompareCatalogRelation` in this file, add the comparison for indices part.

saketj replied:
Updated.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -0,0 +1,177 @@
+   * @brief Constructor.
+   **/
+  IndexScheme() {
+  }
```
You can remove this default constructor, which will always be provided by the compiler. But it's up to you.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+   **/
+  IndexProperties(IndexSubBlockDescription *index_sub_block_description,
+                  const InvalidIndexType &invalid_index_type,
+                  const ParseKeyValue *invalid_property_node)
```
Do you need to provide the default values, for example, `InvalidIndexType::kNone` for `invalid_index_type`, and `nullptr` for `invalid_property_node`?

saketj replied:
Added default values and updated the constructors.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+   * @param invalid_property_node A pointer to an invalid ParseKeyValue node, if any.
+   **/
+  IndexProperties(IndexSubBlockDescription *index_sub_block_description,
+                  const InvalidIndexType &invalid_index_type,
```
Remove `&`, as the strongly typed enum class is still an integer.

saketj replied:
Removed

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+      case InvalidIndexType::kProjectedCountIsNegative:
+        return "projected_element_count cannot be negative";
+      default:
+        return "unknown reason";
```
I suggest to capitalize the first letter in all returned strings.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+  }
+
+ private:
+  valid_property_map_type valid_property_map_;  // a map of index property constants and their associated type.
```
Move the comment in a new line above the member.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+
+  /**
+   * @brief Constructor.
+   *
```
Add comments about `@note The constructor takes ownership of \c index_sub_block_description.`

saketj replied:
Added.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+   *
+   * @return The parse node.
+   **/
+  virtual const ParseKeyValue* getInvalidPropertyNode() const = 0;
```
This method does not need to be `virtual`, let alone `pure virtual`: just return `invalid_property_node_` is fine.

Then remove the overridden versions as well.

saketj replied:
Updated and removed overridden versions.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+   *
+   * @return True if index description is valid, false otherwise.
+   **/
+  virtual bool isIndexDescriptionValid() const = 0;
```
This method does not need to be `virtual`, let alone `pure virtual`: just compare `invalid_index_type_` with `InvalidIndexType::kNone` is fine.

Then remove the overridden versions as well.

saketj replied:
Renamed and virtual removed.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+  CSBTreeIndexProperties()
+    : IndexProperties(new IndexSubBlockDescription(),
+                      InvalidIndexType::kNone,
+                      nullptr) {
```
We could remove these two above, given that we have the default values in the `IndexProperties` constructor.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -0,0 +1,412 @@
+                       const int column_number,
+                       PtrList<ParseKeyValue> *key_value_list)
+      : ParseTreeNode(line_number, column_number),
+    key_value_list_(key_value_list) {
```
Add four more whitespace indents.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -195,22 +197,54 @@ class ParseStatementCreateIndex : public ParseStatement {
+     * @param attribute_name_list A list of attributes of the relation
+     *                            on which the index has to be created.
+     *                            If specified as null, then index is created
+     *                            on all the attributes.
```
These lines above should align with the first letter of `attribute_name_list` in Line 220.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -195,22 +197,54 @@ class ParseStatementCreateIndex : public ParseStatement {
-        index_type_(index_type),
-        index_property_list_(index_property_list) {
+        attribute_list_(attribute_list),
+        index_type_(index_type) {
```
You should add two more whitespace indents for lines between 238 and 242.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -276,29 +326,46 @@ class ParseStatementCreateIndex : public ParseStatement {
-    std::unique_ptr<PtrList<ParseKeyValue> > index_property_list_;
+    std::unique_ptr<IndexProperties> index_properties_;
+    // Optional custom properties for the index can be specified during creation.
+    std::unique_ptr<ParseIndexProperties> custom_properties_node_;
```
Do you need to change `custom_properties_node_` later? If not, mark it const by declaring `std::unique_ptr<const ParseIndexProperties> custom_properties_node_;`.

saketj replied:
Done.

zuyu commented on parser/tests/Index.test:
```diff
@@ -37,9 +37,18 @@ CREATE INDEX test ON test
+==
+CREATE INDEX bloomIndex ON test USING BLOOMFILTER
+--
+CreateIndexStatement[index_name=bloomIndex,relation_name=test,index_type=2]
```
For `index_type`, could we print sth like `bloom-filter` instead of the magic number `2`?

saketj replied:
Replaced the magic numbers with descriptive texts.

zuyu commented on parser/tests/Index.test:
```diff
@@ -37,9 +37,18 @@ CREATE INDEX test ON test
+==
+CREATE INDEX csbIndex ON test USING CSBTREE
+--
+CreateIndexStatement[index_name=csbIndex,relation_name=test,index_type=0]
```
Like above, prefer to print `cs-b-tree`, instead of the magic number `0`.

Also, add tests for `SMAIndex`. I know we use it by default, but I'd like to see as a test case as well.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -810,13 +810,45 @@ void ExecutionGenerator::convertCreateIndex(
+  std::vector<const CatalogAttribute*> index_attributes;
+  for (const E::AttributeReferencePtr &attribute : physical_plan->index_attributes()) {
+    const CatalogAttribute *catalog_attribute
+      = input_relation->getAttributeByName(attribute->attribute_name());
```
Add two more whitespace indents.

saketj replied:
Done.

zuyu commented on query_optimizer/ExecutionGenerator.cpp:
```diff
@@ -810,13 +810,45 @@ void ExecutionGenerator::convertCreateIndex(
+    if (input_relation->hasIndexWithDescription(index_description)) {
+      // Check if the given index description already exists in the relation.
+      THROW_SQL_ERROR() << "The relation " << input_relation->getName()
+      << " already defines this index on "<< catalog_attribute->getName();
```
Add four more whitespace indents.

saketj replied:
Done.

zuyu commented:
@saketj You need also to add the university copyright for all existing files, such as `query_optimizer/resolver/Resolver.cpp`, `query_optimizer/strategy/OneToOne.cpp`, and many.

zuyu commented on query_optimizer/tests/logical_generator/Index.test:
```diff
@@ -0,0 +1,41 @@
+--
+ERROR: Bloom Filter index is not yet implemented (1 : 48)
+...INDEX bloomIndex ON test(int_col) USING BLOOMFILTER
+                                           ^
```
For completeness, add a test case for SMAIndex.

zuyu commented on query_optimizer/tests/physical_generator/Index.test:
```diff
@@ -0,0 +1,58 @@
+--
+ERROR: Bloom Filter index is not yet implemented (1 : 48)
+...INDEX bloomIndex ON test(int_col) USING BLOOMFILTER
+                                           ^
```
For completeness, add a test case for `SMAIndex`.

zuyu commented on query_optimizer/tests/resolver/Index.test:
```diff
@@ -0,0 +1,78 @@
+--
+ERROR: Bloom Filter index is not yet implemented (1 : 39)
+CREATE INDEX bloomIndex ON test USING BLOOMFILTER;
+                                      ^
```
For completeness, add a test case for `SMAIndex`.

zuyu commented on relational_operators/CreateIndexOperator.cpp:
```diff
@@ -28,11 +29,13 @@ bool CreateIndexOperator::getAllWorkOrders(WorkOrdersContainer *container,
                                            StorageManager *storage_manager,
                                            const tmb::client_id foreman_client_id,
                                            tmb::MessageBus *bus) {
+  database_->setStatus(CatalogDatabase::Status::kPendingIndexAdditions);
```
No need to do so. Remove this and other related changes.

In fact, the status staff is only introduced for dropping relations in the distributed case where block deletions happen remotely.

saketj replied:
Removed.

zuyu commented on relational_operators/CreateIndexOperator.hpp:
```diff
@@ -74,7 +86,9 @@ class CreateIndexOperator : public RelationalOperator {
   CatalogRelation *relation_;
+  CatalogDatabase *database_;
   const std::string &index_name_;
+  const std::vector<IndexSubBlockDescription> index_descriptions_;
```
Could you use `reference` here?

saketj replied:
I think that using a reference here will not work as the corresponding variable that initializes this, is a stack variable.

saketj commented:
Updated the copyrights and added the tests for SMAIndex too.

zuyu commented on catalog/CatalogRelation.cpp:
```diff
@@ -1,6 +1,8 @@
  *   Copyright 2011-2015 Quickstep Technologies LLC.
  *   Copyright 2015-2016 Pivotal Software, Inc.
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
```
Remove the copyright for this unchanged file.

zuyu commented on catalog/CatalogRelation.hpp:
```diff
@@ -1,6 +1,8 @@
  *   Copyright 2011-2015 Quickstep Technologies LLC.
  *   Copyright 2015 Pivotal Software, Inc.
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
```
Remove this copyright for this unchanged file.

zuyu commented on catalog/IndexScheme.hpp:
```diff
@@ -159,7 +164,7 @@ class IndexScheme {
       index_sub_block_descriptions.emplace_back(index_descriptions[i]);
     }
-    index_map_[index_name] = index_sub_block_descriptions;
+    index_map_.emplace(std::make_pair(index_name, index_sub_block_descriptions));
```
How about `index_map_.emplace(index_name, std::move(index_sub_block_descriptions))`?

saketj replied:
Done.

zuyu commented on catalog/tests/Catalog_unittest.cpp:
```diff
@@ -161,6 +164,19 @@ class CatalogTest : public ::testing::Test {
+
+    ASSERT_EQ((expected.index_scheme_ != nullptr), (checked.index_scheme_ != nullptr));
+
+    if (expected.index_scheme_ != nullptr && checked.index_scheme_ != nullptr) {
```
Change to `if (checked.index_scheme_ != nullptr) {`, as you've checked it before.

saketj replied:
Done.

zuyu commented on parser/ParseIndexProperties.hpp:
```diff
@@ -86,14 +87,18 @@ class IndexProperties {
-  virtual const ParseKeyValue* getInvalidPropertyNode() const = 0;
+  const ParseKeyValue* getInvalidPropertyNode() const {
+    return invalid_property_node_;
+  }
```
Move above two methods after `virtual` ones.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -193,9 +195,9 @@ class ParseStatementCreateIndex : public ParseStatement {
-     *                            on all the attributes.
+     *        on which the index has to be created.
+     *        If specified as null, then index is created
+     *        on all the attributes.
```
I believe you could reduce one line by combining above three lines of comments.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -218,9 +220,9 @@ class ParseStatementCreateIndex : public ParseStatement {
-     *                            on all the attributes.
+     *        on which the index has to be created.
+     *        If specified as null, then index is created
+     *        on all the attributes.
```
I believe you could reduce one line by combining above three lines of comments.

saketj replied:
Done.

zuyu commented on parser/ParseStatement.hpp:
```diff
@@ -324,7 +326,20 @@ class ParseStatementCreateIndex : public ParseStatement {
+      const int index_type_enum_val = std::stoi(index_type_->value());
+      switch (index_type_enum_val) {
+        case IndexSubBlockType::kCSBTree:
+          inline_field_values->push_back("csb_tree");
```
Change to `cs_b_tree`.

saketj replied:
Fixed.

zuyu commented:
@saketj Once you address new minor comments, I'll merge this PR. Thanks!

saketj commented:
@zuyu I have applied these suggested edits. Thanks!

zuyu commented:
LGTM!
